### ISSUE TITLE
dbus-broker should pick up environment variables

### ISSUE THREAD

**Issue by damianatorrpm** [bus1/dbus-broker#219]
dbus-broker should understand runtime changes to XDG_DATA_DIRS and start services accordingly when XDG_DATA_DIRS is set somewhere. 

Somewhat related:
https://github.com/flatpak/flatpak/issues/3342

**Comment by dvdhrm** [comment]
I do not understand. What do you mean with "runtime changes to XDG_DATA_DIRS"? Can you explain what exactly is not working?

**Comment by damianatorrpm** [comment]
@dvdhrm flatpak/flatpak#3342 explains it as well.
dbus-broker should be able to start service files even if the XDG_DATA_DIRS environment variable is changed during session runtime (or set on a terminal).
`XDG_DATA_DIRS=$XDG_DATA_DIRS:$AND_NEW_DIR /path/to/dbus_activatable_app_that_needs_and_new_dir_where_service_file_is_at`

**Comment by dvdhrm** [comment]
Lets keep the discussion on the flatpak bug. I subscribed to the github-issue and will help figuring this out. I do not see any wrongdoing of dbus-broker so far, though.

**Comment by damianatorrpm** [comment]
What I see as the incorrect behavior of the dbus-broker is that XDG_DATA_DIRS=$XDG_DATA_DIRS:$SHINY_NEW_DIR /path/to/dbus_activatable_app_that_needs_shiny_new_dir_where_service_file_is_located 
will fail to stat the dbus_activable_app as it does not pick up $SHINE_NEW_DIR as a directory to look inside for service files

**Comment by dvdhrm** [comment]
Yeah, that's what I meant. Just because you adjust `XDG_DATA_DIRS` on your terminal does not mean every application's copy of this variable is adjusted as well. So if you set `XDG_DATA_DIRS` before starting dbus-broker (and make sure to set it in its environment), then it should correctly read them. Note that `./dbus-1/services/` is appended to every directory listed in there.

**Comment by damianatorrpm** [comment]
> Yeah, that's what I meant. Just because you adjust XDG_DATA_DIRS on your terminal does not mean every application's copy of this variable is adjusted as well. So if you set XDG_DATA_DIRS before starting dbus-broker (and make sure to set it in its environment), then it should correctly read them. Note that ./dbus-1/services/ is appended to every directory listed in there.

And flatpak has features like flatpak-installalations (see `man flatpak`) that create a new XDG_DATA_DIR. This is not read until reboot or dbus-broker restart which kicks you out of a wayland session. How do these two parts (dbus-broker and flatpak-installations) work together without reboot?

Probably an environment variable like DBUS_BROKER_DO_READ_NEW_DATA_DIR that makes dbus broker read the new flatpak installation and look also inside it.

**Comment by dvdhrm** [comment]
I don't see what I can do for you here, sorry. The `dbus-broker` implementation follows exactly what `dbus-daemon` does.

Furthermore, I don't know why flatpak expects to be able to add directories to XDG_DATA_DIRS at runtime. Please report this against flatpak. They ought to comment on how this is supposed to work, because from the information I am given I don't see how this model was designed, nor how this is meant to work.

**Comment by dvdhrm** [comment]
I will close this issue here. The linked flatpak issue is, in my opinion, the correct place to discuss how flatpak can improve its configuration handling.

If someone proposes changes to dbus-broker, please feel free to open a new issue, or reopen this one. However, please keep in mind that `dbus-broker-launch` is a compatibility tool to `dbus-daemon`, so any possible deviations either need to be justified or implemented in `dbus-daemon` as well. This is the policy for the compatibility launcher.
